source/git/gitsource.go

```diff
@@ -686,3 +694,17 @@ func tokenScope(remote string) string {
 	}
 	return remote
 }
+
+// getDefaultBranch gets the default branch of a repository using ls-remote
+func getDefaultBranch(ctx context.Context, gitDir, workDir, sshAuthSock, knownHosts string, auth []string, remoteURL string) (string, error) {
+	buf, err := gitWithinDir(ctx, gitDir, workDir, sshAuthSock, knownHosts, auth, "ls-remote", "--symref", remoteURL, "HEAD")
+	if err != nil {
```

Comment:
I think it would be helpful to include a hint on how you would set the branch being used (add `#branchname` to URL). If greenlit, I can make a PR for the hint in error messages.